### DISCUSSION TITLE
Size comparison with react-use

### DISCUSSION THREAD

**Discussion by o-alexandrov** [react-hookz/web#123] [Ideas]
If you compare [react-hookz/web](https://bundlephobia.com/package/@react-hookz/web@2.0.0) with [react-use](https://bundlephobia.com/package/react-use@17.2.4) on BundlePhobia (click the library names to open BundlePhobia).

Then, if you look at individual imports, you will see the majority of hooks increased in size, some of them are twice bigger in this library.

I haven't analyzed the reasons, but one possibility is:
- `react-use` hooks don't support SSR most of the time, whereas `react-hookz/web` seems to attempt to support SSR

For most of the users, supporting SSR is actually useless (because they don't do server-side rendering) and, therefore, harmful based on this library's increased size.
Please consider to come up with a way to have an ability to remove SSR support with dead code removal:
- for example, by wrapping SSR-related logic in an imaginary `process.env.SSR` environment variable, so the users could remove it by assigning `false` during the build time.

EDIT:
- SSR related logic is treeshakeable in `@react-hookz/web` if the following text is replaced during the build-time with `"object"` string
  - `typeof window`
  - `typeof navigator`
  - `typeof document`

**Comment by xobotyi** [comment]
And its not🙃
SSR compatibility, in most cases, is just 2-3 extra clauses, nothing more.

The reason is difference in features and optimisations. I think i can squeeze some bytes from the files, but it wont be a drastic change. `react-hookz` reuses more own hooks than `react-use`

**Reply by xobotyi** [reply]
~~Also, `react-hookz` has lots of comments over the code that is not stripped in distribution bundle, which noticeably increases bundle size.
Plus it has JSDoc comments for every hook, `react-use` do not have it at all.~~

Forget it, bundlephobia builds package for production, thus comments are stripped.

**Reply by xobotyi** [reply]
And one more 'also' is the specificity of bundlephobia's "individual" imports.

It does not count reuse of modules, if you sum all separate hooks - it wont result the value that displayed on very top (whole package size).

**Reply by o-alexandrov** [reply]
Comments (including JSDoc) have nothing to do with the bundle size on Bundlephobia, as the pkgs are being built for production (removing comments by default) by Bundlephobia when the size is analyzed.

Regarding features & optimizations that cause the increase in size, I see hooks like `useNetworkState` are bigger due to an unnecessary complication w/ `useSafeState`, whereas everything else is identical to react-use.
- I won't check more hooks, but seems like there are hooks that not only are identical in features, but actually less optimized

And regarding SSR:
- you don't want to introduce ability to remove SSR features as dead code for users who don't use SSR; sounds unreasonable

I just hope that this discussion would make you to start to take care about the size differences more.

**Reply by xobotyi** [reply]
@o-alexandrov 
- `useNetworkState` has async operations (window event handling) thus `useSafeState` use is justified, as it avoids state set of unmounted component and therefore prevents memory leaks. So you literally asking to write less reliable code for the sake of bundle size reduction.
- only 9 hooks, for now, has the browser-check code, the gain from such change will be so miserable that dont even worth of time that would be spent implementing it. Not saying that this (57 bytes):
  ```ts
  const titleRef = useRef(isBrowser ? document.title : '');
  ```
  will become this in sources:
  ```ts
  let docTitle: string;
  if (process.env.SSR === 'false') {
    docTitle = document.title;
  } else {
    docTitle = isBrowser ? document.title : '';
  }
  
  const titleRef = useRef(docTitle);
  ```
  and this in `SSR=false` environment build (77 bytes):
  ```ts
  let docTitle;
  docTitle = document.title;
  const titleRef = useRef(docTitle);
  ```
  So the question: does it worth it?

**Reply by o-alexandrov** [reply]
if you use [Terser's playground](https://try.terser.org/), you would actually end up in the second case (with SSR dead code removal) as:
```ts
const titleRef = useRef(document.title);
```

But, I was probably unclear. Even those, who do use SSR would benefit from the `process.env.SSR`, as when they build for SSR, they could remove all browser-related code.

**Reply by o-alexandrov** [reply]
> `useNetworkState` has async operations (window event handling) thus useSafeState use is justified, as it avoids state set of unmounted component and therefore prevents memory leaks. So you literally asking to write less reliable code for the sake of bundle size reduction.

In this case, it might be subjective a bit, but has nothing to do with the reliability and memory leaks, as in the worst case:
- it'd be called only once after the component is unmounted

But to be clear, those async operations happen so quickly, what you are telling is:
- network state would change, when the component is being unmounted

Sounds unreasonable, doesn't it?

**Reply by o-alexandrov** [reply]
Btw, that's how isBrowser check could be removed during build time.
And everything would still work for those, who don't define `process.env.SSR` during build-time.
For example [here](https://github.com/react-hookz/web/blob/master/src/util/const.ts#L3):
```ts
export const isBrowser =
  typeof window !== 'undefined' &&
  typeof navigator !== 'undefined' &&
  typeof document !== 'undefined';
```

if it's changed to:
```ts
export const isBrowser = !process.env.SSR ||
  (typeof window !== 'undefined' &&
  typeof navigator !== 'undefined' &&
  typeof document !== 'undefined');
```

then, these checks would have been removed during build-time.

**Reply by o-alexandrov** [reply]
actually, after thinking about it.
please ignore everything on SSR I mentioned here, I will just replace:
- `typeof window` 
- `typeof navigator`
- and `typeof document`

with `"object"` during the build-time and the dead code removal in this library would work for my usecase

**Reply by xobotyi** [reply]
> if you use Terser's playground

Obviously not everyone use terser.

> But, I was probably unclear. Even those, who do use SSR would benefit from the process.env.SSR, as when they build for SSR, they could remove all browser-related code.

AFAIK, SSR uses the same bundle for server side and for client side, thats the point of this technology (or i'm missing something).

>But to be clear, those async operations happen so quickly, what you are telling is:
> - network state would change, when the component is being unmounted
> Sounds unreasonable, doesn't it?

It does not, it seems that you dont count slow devices and the murphy's law😑

**Reply by o-alexandrov** [reply]
> AFAIK, SSR uses the same bundle for server side and for client side, thats the point of this technology (or i'm missing something).

No, not always. Some bundle for the SSR separately, removing unused parts to minimize memory consumption and improve performance.
- even if I did it, then there are many others who do it also

Thank you for the discussion. I ended up improving the processing of `@react-hookz/web` for my use case. (modified original message at the top.)

**Reply by xobotyi** [reply]
So what the savings are? Guess something around 100 bytes ungzipped (but minified, ofc) 🤔

**Reply by o-alexandrov** [reply]
Down 1 gripped KB in the minified code (without comments).
From 250 to 249 gzipped KB.

1KB is 30% of preact :D 
But the size reduction is also based on other libraries that depend on window and related checks.
